packages/rest/src/router/metadata.ts

```diff
@@ -261,6 +264,65 @@ export function operation(verb: string, path: string, spec?: OperationObject) {
 
 const paramDecoratorStyle = Symbol('ParamDecoratorStyle');
 
+/**
+ * Check if the given type is `stream.Readable` or a subclasses of
+ * `stream.Readable`
+ * @param type JavaScript type function
+ */
+function isReadableStream(type: Object): boolean {
+  if (typeof type !== 'function') return false;
+  if (type === Stream.Readable) return true;
+  return isReadableStream(Object.getPrototypeOf(type));
+}
+
+/**
+ * Get openapi type name for a JavaScript type
+ * @param type JavaScript type
+ */
+function getNonBodyParamType(type: Function): ParameterType {
+  if (type === String) {
+    return 'string';
+  } else if (type === Number) {
+    return 'number';
+  } else if (type === Boolean) {
+    return 'boolean';
+  } else if (type === Array) {
+    return 'array';
+  } else if (isReadableStream(type)) {
+    return 'file';
+  }
+  return 'string';
+}
+
+/**
+ * Get openapi schema for a JavaScript type for a body parameter
+ * @param type JavaScript type
+ */
+function getSchemaForBodyParam(type: Function): SchemaObject {
+  const schema: SchemaObject = {};
+  let typeName;
+  if (type === String) {
+    typeName = 'string';
+  } else if (type === Number) {
+    typeName = 'number';
+  } else if (type === Boolean) {
+    typeName = 'boolean';
+  } else if (type === Array) {
+    // item type cannot be inspected
+    typeName = 'array';
```

Comment:
Wouldn't you be able to leverage the info provided by `@property.array` and instead take the unknown type in the default case and then check for `array: true` to determine that it's an array of that type?

Reply:
A part of Kyu's current work would be to allow the decorators to take classes instead of SchemaObjects and instead build that metadata into the JSON Schema/OpenAPI/whatever we need. This means that you'd be able to leverage that metadata on incoming requests to link back to the appropriate object type, wouldn't you?

Reply:
It seems like this replaces the need. There is a test case for this as well (the last one).

Reply:
This only seems like it's purpose is to handle the types of incoming items (based on their JavaScript types). I'm referring to the idea that your work will help build a metadata representation of the models that can then be used to build the schema. All of that data should be usable in both directions, meaning we can construct the OpenAPI spec from the back end (for the consumer), and I *think* we should be able to read requests from the front-end and automatically associate them with their corresponding models using the OpenAPI spec that's associated with the selected route (route -> spec -> `$ref` -> metadata for the model object -> model class constructor).

Though, I guess you still need to validate the object after construction. Maybe I'm just a bit lost here.

Reply:
> Wouldn't you be able to leverage the info provided by @property.array and instead take the unknown type in the default case and then check for array: true to determine that it's an array of that type?

The use case is very different. The PR allows a developer to decorate array parameters of a controller method. The array parameter is not a property of a model. Both decorators infer similar metadata, but one is for parameters and the other for properties.

> A part of Kyu's current work would be to allow the decorators to take classes instead of SchemaObjects and instead build that metadata into the JSON Schema/OpenAPI/whatever we need. This means that you'd be able to leverage that metadata on incoming requests to link back to the appropriate class definition*, wouldn't you?

Yes. That's why the PR generates `$ref` for the complex type so that `MyType` (class) can be inspected and mapped a schema object. The parameter definition then references it.